crates/store/re_video/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+//! Video decoding library.
+//!
+//! The entry point is [`load_mp4`], which produces an instance of [`VideoData`].
+
+mod mp4;
+pub use mp4::load_mp4;
+use ordered_float::OrderedFloat;
+
+/// Decoded video data.
+#[derive(Clone)]
+pub struct VideoData {
+    pub config: Config,
+
+    /// Duration of the video, in milliseconds.
+    pub duration: TimeMs,
+
+    /// We split video into segments, each beginning with a key frame,
+    /// followed by any number of delta frames.
+    pub segments: Vec<Segment>,
+
+    /// This array stores all data used by samples.
+    pub data: Vec<u8>,
+}
+
+/// A segment of a video.
+#[derive(Clone)]
+pub struct Segment {
+    /// Time of the first sample in this segment, in milliseconds.
+    pub timestamp: TimeMs,
+
+    /// List of samples contained in this segment.
+    /// At least one sample per segment is guaranteed,
+    /// and the first sample is always a key frame.
+    pub samples: Vec<Sample>,
+}
+
+/// A single sample in a video.
+#[derive(Debug, Clone)]
+pub struct Sample {
+    /// Time at which this sample appears, in milliseconds.
+    pub timestamp: TimeMs,
+
+    /// Duration of the sample, in milliseconds.
+    pub duration: TimeMs,
+
+    /// Offset into [`VideoData::data`]
+    pub byte_offset: u32,
+
+    /// Length of sample starting at [`Sample::byte_offset`].
+    pub byte_length: u32,
+}
+
+/// Configuration of a video.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// String used to identify the codec and some of its configuration.
+    pub codec: String,
+
+    /// Codec-specific configuration.
+    pub description: Vec<u8>,
+
+    /// Natural height of the video.
+    pub coded_height: u16,
+
+    /// Natural width of the video.
+    pub coded_width: u16,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TimeMs(OrderedFloat<f64>);
```

Comment:
What's the purpose of using this over `std::time::Duration`?

We generally try to keep things Rust idiomatic (Duration) and/or SI (seconds)

Reply:
I just felt like `Duration` was a bit more annoying to use. It has `as_millis(&self) -> u128` or `as_secs_f64(&self) -> f64`, both of which require a bit more conversion. All the web APIs also assume milliseconds, so knowing that everything is always in milliseconds is convenient